### ISSUE TITLE
[image_picker] Image_Picker does not open gallery app and selecting images from other apps (e.g. OneDrive) does not work

### ISSUE THREAD

**Issue by alex0711** [flutter/flutter#75020]
## Steps to Reproduce

Open the image picker:
var picker = ImagePicker();
var image = await picker.getImage(source: ImageSource.gallery);

**Expected results:** 
a) the gallery app opens
b) if it is possible to select images from different, third party apps like OneDrive, Google Drive, etc., it should work

**Actual results:**
a). The file manager opens and not the gallery App
b). Selecting images from OneDrive (open the drawer, select OneDrive, select an image) does not work => see issue #73099

**Idea**
I expected from the enum ImageSource.gallery that actually the gallery app opens.
To achieve this, we would need to change [launchPickImageFromGallery](https://github.com/flutter/plugins/blob/07dade429d22ca7d8659db2d2fde2fde3acfaae6/packages/image_picker/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/ImagePickerDelegate.java#L334) to use the following Intent instead:

`Intent pickImageIntent = new Intent(Intent.ACTION_PICK, android.provider.MediaStore.Images.Media.EXTERNAL_CONTENT_URI);`

This would ensure the gallery app is opened.

If it is desired for the #pickImage method to also allow selection of images from OneDrive (and others), then it could be changed to:
`   Intent pickImageIntent = new Intent(Intent.ACTION_OPEN_DOCUMENT);
    pickImageIntent.addCategory(Intent.CATEGORY_OPENABLE);
    pickImageIntent.setType("image/*");
`
Then we'd just need some more logic in FileUtils#getPathFromUri: currently the Intents data is always expected to be the file URI. That may not always be the case, it may be necessary to get the uri by invoking:
`val uri = intent.getParcelableExtra<Parcelable>(Intent.EXTRA_STREAM) as Uri`

@cyanglaz: I don't know what's the intended behavior - show the gallery app or the file manager which enables selection from OneDrive and others? I guess it would help to leave the decision which app to open to the caller: ImageSource.gallery could open the gallery app (anything else is not really expected judgin from the enum name) and a new enum ImageSource.fileManager could open the file manager.

Flutter doctor output:
alex@S3409:~/Development/xxx/xxx$ fvm flutter doctor -v
[✓] Flutter (Channel stable, 1.22.5, on Linux, locale en_US.UTF-8)
    • Flutter version 1.22.5 at /home/alex/fvm/versions/stable
    • Framework revision 7891006299 (7 weeks ago), 2020-12-10 11:54:40 -0800
    • Engine revision ae90085a84
    • Dart version 2.10.4

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /home/alex/Android/Sdk
    • Platform android-30, build-tools 30.0.2
    • Java binary at: /opt/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[!] Android Studio
    • Android Studio at /opt/android-studio
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.52.1)
    • VS Code at /usr/share/code
    • Flutter extension version 3.19.0

[✓] Connected device (1 available)
    • SM A520F (mobile) • android-arm64 • Android 8.0.0 (API 26)

**Comment by darshankawar** [comment]
@alex0711 
I tried the official `image_picker` plugin example and using source as `ImageSource.gallery`, the app properly opened gallery for me:

<img width="541" alt="Screenshot 2021-02-01 at 1 48 47 PM" src="https://user-images.githubusercontent.com/67046386/106432233-7113ea00-6494-11eb-816d-8865d1e4d7e1.png">

Also showed various other available sources from the phone to choose images from:

<img width="423" alt="Screenshot 2021-02-01 at 1 49 16 PM" src="https://user-images.githubusercontent.com/67046386/106432281-825cf680-6494-11eb-82f6-252bde610838.png">

<img width="414" alt="Screenshot 2021-02-01 at 1 49 22 PM" src="https://user-images.githubusercontent.com/67046386/106432288-8426ba00-6494-11eb-836c-936f1b21168f.png">

Can you describe as what exactly your issue is and how different is this from https://github.com/flutter/flutter/issues/73099 ?

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 1.22.6, on Mac OS X 10.15.4 19E2269 darwin-x64,
    locale en-IN)
    • Flutter version 1.22.6 at /Users/dhs/documents/Fluttersdk/flutter
    • Framework revision 9b2d32b605 (9 days ago), 2021-01-22 14:36:39 -0800
    • Engine revision 2f0af37152
    • Dart version 2.10.5

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.0)
    • Android SDK at /Users/dhs/Library/Android/sdk
    • Platform android-30, build-tools 30.0.0
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    • CocoaPods version 1.9.3

[!] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.52.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.18.1

[✓] Connected device (2 available)
    • SM A260G (mobile)            • 5200763ebcfa861f • android-arm • Android
      8.1.0 (API 27)
    • AOSP on IA Emulator (mobile) • emulator-5554    • android-x86 • Android 9
      (API 28) (emulator)

! Doctor found issues in 1 category.


```

</details>

**Comment by alex0711** [comment]
@darshankawar  thanks for your reply. Here's how the image_picker with ImageSource.gallery behaves for me:

1) A file manager app is opened which shows recently used images
![file_manager resized](https://user-images.githubusercontent.com/23746778/106482904-3d45be00-64ae-11eb-9fcc-efbaedd4cb69.jpg)

2) In there i can open the drawer and navigate to OneDrive, Google Drive and the Gallery
![fm_drawer resized](https://user-images.githubusercontent.com/23746778/106482929-433b9f00-64ae-11eb-9c9d-df16f1690f8c.jpg)

3) If i select the gallery in the drawer ("Galerie" in the screenshot, german locale), the gallery app opens
![gallery resized](https://user-images.githubusercontent.com/23746778/106482951-49318000-64ae-11eb-8efa-c9061b801b23.jpg)

So for me the app that get's opened by the image_picker is not the gallery app, but some file manager / file selection app.
I can then navigate to other apps (2) but even if i do navigate to the gallery (3) and then select an image there it does not work: The gallery app closes and i'm back at (1). Basically this behavior is issue #73099 just that it seems to apply for all 'further' apps accessed from (2). But also i would have expected that the gallery app is opened by the image_picker, just judging from the enum name.

**Comment by darshankawar** [comment]
@alex0711 
Thanks for the detailed explanation. I do see the SAF UI when I select `gallery` option first.

<img width="372" alt="Screenshot 2021-02-02 at 11 51 22 AM" src="https://user-images.githubusercontent.com/67046386/106560411-007cd400-654d-11eb-8030-9bd038d1650f.png">

But from your quote `and then select an image there it does not work: The gallery app closes and i'm back at (1).`. I selected `gallery` from dropdown and then `Albums` and then a picture from it, took me directly to the app home screen showing selected image from gallery.

<img width="545" alt="Screenshot 2021-02-02 at 11 53 24 AM" src="https://user-images.githubusercontent.com/67046386/106560604-476ac980-654d-11eb-93f5-bdbd4318cdfd.png">

I do agree that selecting `gallery` should show `gallery` source rather than device's inbuilt UI though.

**Comment by alex0711** [comment]
@darshankawar exactly, for me the behavior differs: When i select an image in the gallery i'm back at the SAF UI. I think this could differ due to the different API level and different customized vendor android images.

I just tried it with stock android emulators (API level 26, which is same as my physical device, and API level 29):
1) first take a picture in the emulator
2) then use the image_picker to pick an image
3) open the drawer to navigate to the Google Photos app
4) then select the image taken at step 1
5) and we're not navigating back to the flutter app but we're back at the SAF UI
Same behavior for both API level 26 and 29 in the emulator for me.

So basically SAF UI seems not to be stable with regard to image selection, which is not a problem of SAF UI in general but probably of the Apps that implement SAF capability (like OneDrive, Samsung Gallery, ...) on different devices and versions.

So i'd welcome it if we could make ImageSource.gallery open the default gallery app by default - as mentioned in the opening post, this is just a change of the Intent which should be used.

**Comment by darshankawar** [comment]
Thanks @alex0711 
I see this issue more as a proposal to make `ImageSource.gallery open the default gallery app by default`.

**Comment by ram231** [comment]
I have several apps that opens gallery(Google Drive, Photos, File Manager, Downloads). I usually let users choose which default gallery they chose.

**Comment by alex0711** [comment]
@darshankawar right, that would solve my issues. Also it would be consistent with the iOS behavior: AFAIK the UIImagePickerController used there does not allow to select images provided by other apps, it basically just let's you select the images from your albums in the iOS gallery app.

@ram231: did your users never stumble upon document providers (aka one of the drawer menu items in the SAF UI, e.g. OneDrive, Google Drive, Gallery, ...) that do not work and produce the behavior i'm experiencing on the emulator and on my physical device?

Introducing a third enum value ImageSource.xyz could be used to open the SAF UI instead on Android - though that would make the API a little cumbersome, as then this new enum value would only have different semantics on Android (well, with cross-app frameworks i think these kind of 'issues' will always appear). So i'd prefer to let the app decide which app to open (SAF UI or Gallery app), as the SAF UI just doesn't work for all cases (see related issue #73099) and sometimes it is just intended to directly show the gallery (which is less error prone and saves the user a few taps).

**Comment by xkxeeshankhan** [comment]
I'm facing the similar problem, when selecting an image from Photo or Gallery, it takes user to the `SAF UI`.
version: `image_picker: 0.8.5+1 `

https://user-images.githubusercontent.com/44693638/167921631-bdc93eda-1f95-4728-8275-ef4b82484c13.mp4

**Comment by luvishq** [comment]
@alxflam : Did you find any solution to this ?

**Comment by alxflam** [comment]
@luvishq for testing the issue i applied the fix that i mentioned in the opening post (adapting [launchPickImageFromGallery](https://github.com/flutter/plugins/blob/07dade429d22ca7d8659db2d2fde2fde3acfaae6/packages/image_picker/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/ImagePickerDelegate.java#L334) and then using a custom version of the image_picker package). Anyways i no longer use a phone whose SAF UI behaves weirdly and therefore i no longer experience the issue.

**Comment by thetmyoekhaing** [comment]
> ## Steps to Reproduce
> 
> Open the image picker: var picker = ImagePicker(); var image = await picker.getImage(source: ImageSource.gallery);
> 
> **Expected results:** a) the gallery app opens b) if it is possible to select images from different, third party apps like OneDrive, Google Drive, etc., it should work
> 
> **Actual results:** a). The file manager opens and not the gallery App b). Selecting images from OneDrive (open the drawer, select OneDrive, select an image) does not work => see issue #73099
> 
> **Idea** I expected from the enum ImageSource.gallery that actually the gallery app opens. To achieve this, we would need to change [launchPickImageFromGallery](https://github.com/flutter/plugins/blob/07dade429d22ca7d8659db2d2fde2fde3acfaae6/packages/image_picker/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/ImagePickerDelegate.java#L334) to use the following Intent instead:
> 
> `Intent pickImageIntent = new Intent(Intent.ACTION_PICK, android.provider.MediaStore.Images.Media.EXTERNAL_CONTENT_URI);`
> 
> This would ensure the gallery app is opened.
> 
> If it is desired for the #pickImage method to also allow selection of images from OneDrive (and others), then it could be changed to: ` Intent pickImageIntent = new Intent(Intent.ACTION_OPEN_DOCUMENT); pickImageIntent.addCategory(Intent.CATEGORY_OPENABLE); pickImageIntent.setType("image/*");` Then we'd just need some more logic in FileUtils#getPathFromUri: currently the Intents data is always expected to be the file URI. That may not always be the case, it may be necessary to get the uri by invoking: `val uri = intent.getParcelableExtra<Parcelable>(Intent.EXTRA_STREAM) as Uri`
> 
> @cyanglaz: I don't know what's the intended behavior - show the gallery app or the file manager which enables selection from OneDrive and others? I guess it would help to leave the decision which app to open to the caller: ImageSource.gallery could open the gallery app (anything else is not really expected judging from the enum name) and a new enum ImageSource.fileManager could open the file manager. I could provide a pull request with these changes :)
> 
> Flutter doctor output: alex@S3409:~/Development/xxx/xxx$ fvm flutter doctor -v [✓] Flutter (Channel stable, 1.22.5, on Linux, locale en_US.UTF-8) • Flutter version 1.22.5 at /home/alex/fvm/versions/stable • Framework revision [7891006](https://github.com/flutter/flutter/commit/78910062997c3a836feee883712c241a5fd22983) (7 weeks ago), 2020-12-10 11:54:40 -0800 • Engine revision ae90085a84 • Dart version 2.10.4
> 
> [✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2) • Android SDK at /home/alex/Android/Sdk • Platform android-30, build-tools 30.0.2 • Java binary at: /opt/android-studio/jre/bin/java • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593) • All Android licenses accepted.
> 
> [!] Android Studio • Android Studio at /opt/android-studio ✗ Flutter plugin not installed; this adds Flutter specific functionality. ✗ Dart plugin not installed; this adds Dart specific functionality. • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
> 
> [✓] VS Code (version 1.52.1) • VS Code at /usr/share/code • Flutter extension version 3.19.0
> 
> [✓] Connected device (1 available) • SM A520F (mobile) • android-arm64 • Android 8.0.0 (API 26)

### same here , having difficulties opening gallery :(

**Comment by oligazar** [comment]
ImageSource.gallery still opens SAF UI instead of gallery app. 
1. It's confusing. Why call it .gallery, if it doesn't open gallery?
2. Anyways, there should be a way to specify whether you want SAF or gallery app to be opened

**Comment by JettChen12** [comment]
image_picker1.2.1 continues to experience this issue on older Android versions, specifically Android 11 (reproduced on an OPPO A11).
